### ISSUE TITLE
Invalidating bandwidth levels on DASH adaptation change

### ISSUE THREAD

**Issue by SteUK** [google/ExoPlayer#2336]
Hi,

Not sure whether this is a 'how do I achieve this' question, or a feature request!

I am playing back a DASH stream with multiple adaptation sets (think multiple camera angles).  When I change stream, I want to minimise the delay before the new camera starts playing back.  Essentially I would like to ensure the lowest bitrated representation is streamed in and started.  What I am seeing on adaptation set change is the same bitrate representation is being buffered and then played.  Is there a way to say 'start buffering the lowest bitrated representation, rather than the level currently being enjoyed?'

Hmm, having written that...maybe the answer is a lot simpler.  When I change adaptation sets could I be selective and only put one track index in the tracks array passed to 'MappingTrackSelector.SelectionOverride' ...then later on, when playback has started, add the other higher bitrate tracks?  Or would this then cause a re-buffering?

Thanks!

**Comment by ojw28** [comment]
> Hmm, having written that...maybe the answer is a lot simpler. When I change adaptation sets could I be selective and only put one track index in the tracks array passed to 'MappingTrackSelector.SelectionOverride' ...then later on, when playback has started, add the other higher bitrate tracks? Or would this then cause a re-buffering?

Yes, it would cause re-buffering.

> Is there a way to say 'start buffering the lowest bitrated representation, rather than the level currently being enjoyed?'

There's no trivial way to enable this, but you can do what you want with a bit of work:

- If you look at what the demo app does when it creates its `DefaultTrackSelector` instance, it passes an `AdaptiveVideoTrackSelection.Factory` as an argument to the constructor. Once the set of tracks has been chosen by the selector, it invokes this factory to build a `TrackSelection` containing the individual selected tracks.
- The `TrackSelection` built by the factory is what controls the adaptation between the individual selected tracks. `AdaptiveVideoTrackSelection.Factory` creates instances of `AdaptiveVideoTrackSelection`.
- You can provide your own factory that provides your own implementation of `TrackSelection`. You'd probably want to fork `AdaptiveVideoTrackSelection` into your own codebase and then adapt it to meet your needs. In particular, you can force the selected track to the lowest quality in your forked implementation until `bufferedDurationUs` first exceeds some (presumably relatively small) value, or something along those lines.

**Comment by WeiChungChang** [comment]
Hi @SteUK :
How about your test result?
I gave it a try as per @ojw28 ’s kind suggestion.

Please refer to the patch named “Implement-Slow-start-to-accelerate-adaptationset-switch.zip”; where I tried to customize AdaptiveVideoTrackSelection accordingly.
[Implement-Slow-start-to-accelerate-adaptationset-switch.zip](https://github.com/google/ExoPlayer/files/762958/Implement-Slow-start-to-accelerate-adaptationset-switch.zip)

The name "slow start" comes from GStreamer since it always starts from the representation with lowest bitrate when creating a new stream.

There are 2 considerations when I created it.
1.	A function named getLowestSelectedIndex() is used to select (start from) the representation with lowest bitrate when a new instance of  SlowStartAdaptiveVideoTrackSelection is created. Since slowStartAdaptiveVideoTrackSelection is created only when we have a new setSelectionOverride, it insures a faster transition between adaptationsets.
2.	Since the transition roughly includes 2 stages: 
A: render the first frame.
B: have sufficient buffer.
A value named "slowStartInitializationPeriod" is set to equal to 
“public static final int DEFAULT_BUFFER_FOR_PLAYBACK_MS = 2500” of DefaultLoadControl.java. It ensures us to keep on downloading the representation with lowest bitrate until we have had sufficient buffering data to drive playback starts.

The patch below is the way I launch the new trackselection.
[laucher.zip](https://github.com/google/ExoPlayer/files/763023/laucher.zip)
 I tested it by the steps below:
1.	Install http://www.rejetto.com/hfs/ on my laptop.
2.	Connect the mobile phone and the server (my laptop) to the same WIFI.
3.	Rewrite the test MPD to simulate 2 video adaptation sets(as the multi-view case you described).
4.	Switch the video tracks and record how fast the transition is; compare the effects of original AdaptiveVideoTrackSelection with the new SlowStartAdaptiveVideoTrackSelection

Please refer to the modified MPD below for testing also.
[car-20120827-manifest_dual_video_adaptationset.zip](https://github.com/google/ExoPlayer/files/762947/car-20120827-manifest_dual_video_adaptationset.zip)

The result shows the in average under connection status of network speed is about **17.71 Mbps**. It seems to make the transition faster but is still not seamless.
In conclusion, it makes **26.3%** speedup to render the first frame & **87.35%** to start playback (after having sufficient buffer)
![test_result_0209](https://cloud.githubusercontent.com/assets/14846473/22775003/aa45946a-eee4-11e6-80c1-e895b3ecbf21.png)

What’s your next step for accelerating it? 
It is an interesting problem. Once you have made a scheme to improve it, I will be glad if you could share it.

Thanks.

**Comment by IA-Ben** [comment]
Hi @WeiChungChang 
Thank you very much for this. @SteUK is away for the next week but will look at this once he's back.
Best, Ben
